pipeline/include/cvs/pipeline/imodule.hpp

```diff
@@ -1,33 +1,29 @@
 #pragma once
 
 #include <boost/dll/shared_library.hpp>
-#include <cvs/common/configuration.hpp>
 #include <cvs/common/factory.hpp>
+#include <cvs/pipeline/version.hpp>
 
 #include <memory>
 
 namespace cvs::pipeline {
 
 class IModule {
  public:
-  virtual ~IModule()                        = default;
-  virtual std::string name() const          = 0;
-  virtual int         version() const       = 0;
-  virtual void        registerTypes() const = 0;
-  virtual bool        checkCompatibility() const {
-    return common::Factory::libVersion == common::Factory::kVersion && libVersion == kVersion;
-  }
+  virtual ~IModule()                                                            = default;
+  virtual std::string name() const                                              = 0;
+  virtual int         version() const                                           = 0;
+  virtual void        registerTypes(cvs::common::FactoryPtr<std::string>) const = 0;
```

Comment:
Вроде здесь должно быть достаточно `const cvs::common::Factory<std::string> &`

Reply:
Конкретно в этом месте предполагалось, что модуль может скопировать фабрику и держать ее. Напрмиер чтобы выполнить какие-либо действия в деструкторе (разрегистрировать что-то, правда фабрика такого пока не умеет). Отсюда и аргуметн такой.

Reply:
Не уверен в реалистичности такого юзкейса, но ок, пусть будет. По крайней мере стоит сделать аргумент константной ссылкой (модуль может копировать с разделением владения, а может не копировать).

Reply:
Дык из соображений такого юзкейса и сделано. Мол явно либо надо сделать move либо отдать копию. А модуль уже сохранит у себя.

Маерс вроде про что-то такое писал.

Нужно определиться как и что могут делать модули. Через это будет понятнее как писать.

Reply:
Я понимаю, что текущий тип — хороший способ показать, что модули разделяют владение фабрикой. Однако, если это нужно не всегда / не всем модулям, то лучше подойдёт `const cvs::common::FactoryPtr<std::string> &`

Reply:
Тут противопоставляется следующее:
лишнее копирование с инкрементом и декрементом счетчика ссылок
и
указание на то, что указатель на фабрику могут запомнить.

В целом принебречь что первым что вторым - не критично. Но я склоняюсь принебречь первым. @efimov-evgeniy на твуй взгляд как лучше?

Reply:
Само по себе использование `shared_ptr` в типе аргумента уже указывает на то, что его могут запомнить.

Reply:
[Take a const shared_ptr<widget>& parameter to express that it might retain a reference count to the object](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#r36-take-a-const-shared_ptrwidget-parameter-to-express-that-it-might-retain-a-reference-count-to-the-object-)

Reply:
в таком случае надо переписывать в др местах.

Reply:
Если подразумевается, что в месте вызова функция `registerTypes` не "поглощает" указатель, а только лишь имеет возможность скопировать объект `shared_ptr`, то имхо логично использовать `const &`.